kythe/go/extractors/cmd/rust/rust_project_to_kzip.go

```diff
@@ -0,0 +1,552 @@
+	for _, crate := range crates {
+		transitiveDepsForCrate := transitiveDepsForCrate(crate, crateDeps)
+		transitiveDeps[crate.CrateId] = transitiveDepsForCrate
+	}
+	return transitiveDeps
+}
+
+// getSourceDirs returns the set of included and excluded directories for all crates
+// Additionally, we add the directory of the root module to the set of included directories.
+// It returns a data structure containing the source directories and dependent directories, paired by include and exclude.
+// Note: the returned source directories are NOT transitive
+func getSourceDirs(crates []crate) map[crateId]source {
+	perCrateSourceDirs := make(map[crateId]source)
+	for _, crate := range crates {
+		currentSourceDirs := source{
+			IncludeDirs: []string{},
+			ExcludeDirs: []string{},
+		}
+		currentSourceDirs.IncludeDirs = append(currentSourceDirs.IncludeDirs, crate.Source.IncludeDirs...)
+		currentSourceDirs.ExcludeDirs = append(currentSourceDirs.ExcludeDirs, crate.Source.ExcludeDirs...)
+
+		// Add the dir of the root module for the crate, which is implicitly included
+		currentSourceDirs.IncludeDirs = append(currentSourceDirs.IncludeDirs, filepath.Dir(crate.RootModule))
+
+		perCrateSourceDirs[crate.CrateId] = currentSourceDirs
+	}
+
+	return perCrateSourceDirs
+}
+
+// isSubdirectory checks if the path `sub` is a subdirectory of `parent`.
+func isSubdirectory(parent, sub string) bool {
+	return !strings.HasPrefix(sub, parent)
+}
+
+// whether we should include this file in the Compilation Unit based on the inclusions/exclusions in `sourceDirs`
+func shouldIncludeFile(info os.FileInfo, path string, sourceDirs source, projectRoot string) (bool, error) {
+	if info.IsDir() {
+		// check that this file isn't in one of the dirs excluded by this crate
+		for _, excludeDir := range sourceDirs.ExcludeDirs {
+			if isSubdirectory(filepath.Join(projectRoot, excludeDir), filepath.Dir(path)) {
+				return false, filepath.SkipDir
+			}
+		}
+		return false, nil
+	}
+
+	// only index .rs files for now. It's possible that some rust files depend on non rust files, like if they use include_bytes,
+	// TODO(wittrock): figure out a better scheme for including all actually depended-on files and still exclude files like .deps and .rmeta
+	if filepath.Ext(path) != ".rs" {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+type collectCrateSourcesFunc func(ctx context.Context, e *extractor, sourceDirs source) ([]string, []*apb.CompilationUnit_FileInput, error)
+
+// getSourceFiles walks the file system based on the provided sourceDirs,
+// identifies relevant source files (currently only .rs files), adds them to the
+// kzip archive, and returns lists of their relative paths and FileInput protos.
+func collectCrateSourcesImpl(ctx context.Context, e *extractor, sourceDirs source) ([]string, []*apb.CompilationUnit_FileInput, error) {
+	var sourceFiles []string
+	var requiredInputs []*apb.CompilationUnit_FileInput
+
+	// cache required inputs for each include_dir
```

Comment:
it's possible to have different crates sharing the same include_dir but different exclude_dirs, in which case your cache is going to give the wrong answer. exclude_dirs could be part of the cache key.

Alternatively: collectCrateSourcesImpl could totally ignore exclude_dirs and include all transitive files! The extra files would be available in the VFS, but rust-analyzer would never look at them because `exclude_dirs` would still be set in `rust-project.json`. The limitation of this approach is if there's a huge file tree that's supposed to be excluded from all crates, we'll blow up the kzip size unneccesarily.

Up to you what to do here, TODO is also fine

Reply:
Two problems in computer science: naming, cache invalidation, and off-by-one errors. 

Great catch. What a pain it is to use a struct with a slice member as a key to a map in Go! But I've done it.